readline/buffer.go

```diff
@@ -121,114 +163,258 @@ func (b *Buffer) MoveToEnd() {
 			remainder := b.Size() % b.LineWidth
 			fmt.Printf(CursorBOL + cursorRightN(len(b.Prompt.prompt())+remainder))
 		} else {
-			fmt.Print(cursorRightN(b.Size() - b.Pos))
+			fmt.Print(cursorRightN(b.Size() - b.DisplayPos))
 		}
 
-		b.Pos = b.Size()
+		b.Pos = b.Buf.Size()
+		b.DisplayPos = b.Size()
 	}
 }
 
 func (b *Buffer) Size() int {
-	return b.Buf.Size()
+	sum := 0
+	for i := 0; i < b.Buf.Size(); i++ {
+		if e, ok := b.Buf.Get(i); ok {
+			if r, ok := e.(rune); ok {
+				sum += runewidth.RuneWidth(r)
+			}
+		}
+	}
+
+	return sum
 }
 
 func (b *Buffer) Add(r rune) {
+	rLength := runewidth.RuneWidth(r)
+
 	if b.Pos == b.Buf.Size() {
-		fmt.Printf("%c", r)
+
+		b.DisplayPos += rLength
+
+		if b.Pos > 0 {
+
+			if b.DisplayPos%b.LineWidth == 0 {
+				fmt.Printf("%c", r)
+				fmt.Printf("\n%s", b.Prompt.AltPrompt)
+				b.LineFlags.Add(false)
+
+			} else if b.DisplayPos%b.LineWidth < (b.DisplayPos-rLength)%b.LineWidth {
```

Comment:
This probably deserves a comment saying something along the lines of "double width char crosses the LineWidth boundary"